### DISCUSSION TITLE
Audio for py5

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#282] [Show and tell]
I'm starting to read this: https://k3no.medium.com/integrating-pyaudio-pysimplegui-5fa342b1e113

Update: this could be useful https://www.henryschmale.org/2021/01/07/pygame-linein-audio-viz.html

It looks very promising. Despite my terrible mental model about anything sound-stream-like (so I have asked a sound-savvy friend to join me on these experiments, hopefully soon), I have already tried my hand at the "blocking example"

![image](https://user-images.githubusercontent.com/3694604/230776810-8864e2b8-d312-411d-8c26-d648f9f6543b.png)

Next I should have a look at the "callback"/"non-blocking"  example and the other PySimpleGUI integration parts to get inspiration for more py5 experiments.

If you are reading this and you would like to try your hand at sound-viz examples with py5, I'd be delighted to see your experiments.

**Comment by hx2A** [comment]
Interesting. I don't know much about sound in Python. This library you found looks like a good find. What audio formats does it support? The documentation is a bit sparse. From what I've seen, Python audio libraries are either simple + don't support mp3s or are complicated + support mp3s.

py5 has `launch_thread()` and `launch_repeating_thread()` methods, which might be useful to you if you don't want to go the callback/non blocking route but want to get the code out of your `draw()` method.

**Reply by villares** [reply]
You are quite right, mp3 is hard. PyAudio doesn't know mp3... Maybe I should have a look at PyDub for mp3s: https://stackoverflow.com/a/53633178/19771 or https://dev.to/mathewthe2/playing-mp3-files-in-python-with-pydub-and-pyaudio-579i

Yeah, cool, I was going to try `launch_repeating_thread()` at some point!

**Reply by villares** [reply]
```python
import py5
import pyaudio
import numpy as np

"""Simple Blocking Stream PyAudio"""
CHUNK = 128  # Samples: 1024,  512, 256, 128
RATE = 44100  # Equivalent to Human Hearing at around 40 kHz
INTERVAL = 0.10  # Sampling Interval in Seconds

p = pyaudio.PyAudio()
stream = p.open(format=pyaudio.paInt16,
                channels=1,
                rate=RATE,
                input=True,
                frames_per_buffer=CHUNK)
sample_size = int(INTERVAL*RATE/CHUNK)
data = []

def setup():
    global sw
    py5.size(600, 600)
    py5.no_fill()
    py5.color_mode(py5.HSB)
    sw = py5.width / sample_size
    py5.launch_repeating_thread(listen, 'listen')

def draw():
    py5.background(240)
    for i, v in enumerate(data):
        py5.stroke((64 + v / 4) % 255, 200, 200)
        py5.circle(i * sw, py5.height / 2, 64 + v / 4)

def listen():
    new_data = []
    for i in range(sample_size):  # STREAM INTERVAL
        sound_data = np.frombuffer(stream.read(CHUNK), dtype=np.int16)
        new_data.append(np.amax(sound_data))
    data[:] = new_data
    
def exiting():
    stream.stop_stream()
    stream.close()
    p.terminate()

py5.run_sketch()
```
:smile_cat:

**Reply by hx2A** [reply]
I would expect the frame rate for this example to be better than the previous one. I like your code here, it is simple and concise. Good use of `data[:]`, that is a good Python technique to use. You might be able to use a numpy array instead of a list though. Isn't it going to be the same size every time?
